### PR TITLE
[SITES-10211] GuestUIFrame resize detection support

### PR DESCRIPTION
**Pull request by zetlen**
## Description

**Based against https://github.com/adobe/uix-sdk/pull/21 which must be merged before this can be reviewed.**

- `@adobe/uix-core`:
  - Refactor some types to allow typed event subscriptions in `Port` and `Guest` objects
  - Add a simple callback throttling utility for later use with resize observers
- `@adobe/uix-guest`:
  - Intake refactored types from `uix-core`
  - When an extension intends to display UI and therefore uses `attach()` instead of `register()`, the UIGuest object watches the document for size changes and sends them to the host object via an `"iframeresize"` event
- `@adobe/uix-host`:
  - Intake refactored types from `uix-core`
  - `host.provide()` now merges method namespaces (two-level-deep merge) instead of overwriting them (one-level-deep merge)
  - `port.attachUI()` now provides the UI guest with a callback for emitting the `"iframeresize"` event on the host side
  - `port.attachFrame()` now accepts an additional set of methods to merge and send down to new frame
- `@adobe/uix-react`:
  - Intake refactored types from `uix-core`
  - `GuestUIFrame` refactored to support detecting and responding to the inner document resizing

## Related Issue

[SITES-10256](https://jira.corp.adobe.com/browse/SITES-10256)

## Motivation and Context

Includes necessary changes to support the pull request https://git.corp.adobe.com/CQ/aem-headless-ui-commons/pull/316 for https://jira.corp.adobe.com/browse/SITES-10256.

To resize modals which display GuestUIFrames, it's necessary for guest objects to listen for document size changes, and send new dimensions to the host in a separate event. This seems like it ought to be a part of SDK functionality, but there wasn't an obvious channel to pass this data; since it's going to be built in, it doesn't seem appropriate to take up part of the public API methods namespace.

To add this data channel, we had to open up the Guest and Port objects to be able to exchange arbitrary events. Then, to collect that data, we had to make guest objects subscribe to resize events when they are connected, and pass those dimension values up to the host.

Finally, we needed to handle the very large number of resize events that can fire quickly, which we did by implementing a throttle function and throttling resize event callbacks.

## How Has This Been Tested?

Updated unit tests, tested locally extensively

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/uix-core/src/tunnel/tunnel.test.ts [hunk 1]
```diff
@@ -185,14 +203,18 @@ describe("static Tunnel.toIframe(iframe, options)", () => {
...
       await wait(100);
-      fireEvent(
-        window,
-        new MessageEvent("message", {
-          data: messenger.makeOffered("iframe-test-1"),
-          origin: loadedFrame.src,
-          source: loadedFrame.contentWindow,
-        })
+      window.postMessage(
+        messenger.makeOffered("iframe-test-1"),
```
Comment (vkublytskyi) [inline packages/uix-core/src/tunnel/tunnel.test.ts:207]: Please remove debugging code before merging PR

**Review by vkublytskyi — CHANGES_REQUESTED** [review]

File: packages/uix-core/src/types.ts [hunk 2]
```diff
@@ -164,6 +162,40 @@ export interface HostConnection<T = unknown> {
...
+  HostApi extends Record<string, unknown> = Record<string, unknown>
+> =
+  | GuestConnectionEvent<"hostprovide">
+  | GuestConnectionEvent<"unload">
+  | GuestConnectionEvent<"beforecallhostmethod", HostMethodAddress<HostApi>>
+  | GuestConnectionEvent<
+      "iframeresize",
+      { dimensions: DOMRect; iframe: HTMLIFrameElement }
+    >;
+
```
Comment (zetlen) [inline packages/uix-core/src/types.ts:198]: As part of the `GuestConnection` to `GuestEmitter` refactor, move the event definitions from the `Port` implementation to the `GuestEmitter` interface.

File: packages/uix-core/src/types.ts [hunk 3]
```diff
@@ -219,4 +253,11 @@ export interface Emits<Events extends NamedEvent = NamedEvent>
     type: Type,
     listener: (ev: Extract<Events, { type: Type }>) => unknown
   ): () => void;
+  /**
+   * Same as EventTarget.removeEventListener but typed.
+   */
+  removeEventListener<Type extends Events["type"]>(
+    type: Type,
+    listener: (ev: Extract<Events, { type: Type }>) => unknown
+  ): void;
```
Comment (zetlen) [inline packages/uix-core/src/types.ts:262]: I didn't notice that `removeEventListener` was missing in this interface until I tried to call it from code.

File: packages/uix-core/src/async-throttle.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  intervalMs: number
+) {
+  let throttled = false;
+  return (...args: Parameters<typeof callback>) => {
+    if (throttled) {
+      return;
+    }
+    throttled = true;
+    setTimeout(async () => {
+      await callback(...(args as unknown[]));
```
Comment (kokoc) [inline packages/uix-core/src/async-throttle.ts:21]: Not sure if we need it for our case, but this implementation doesn't support return values from callbacks. 
Also, throttled requests usually somehow identified, so they could be handled differently on client's side

**Review by fe-lix- — APPROVED** [review]
LGTM

File: packages/uix-core/src/async-throttle.ts (hunk 4 above)
Comment (kokoc) [inline packages/uix-core/src/async-throttle.ts:21]: Shouldn't this method call the callback immediately if it's not throttled?

Comment (zetlen) [reply packages/uix-core/src/async-throttle.ts:21]: Yeah, this was buggy code. I have removed it in the latest version, which is much simpler and being pushed shortly

**Review by kokoc — APPROVED** [review]

**Review by vkublytskyi — APPROVED** [review]

**Pull request merged**